pennylane/io/qasm_interpreter.py

```diff
@@ -0,0 +1,305 @@
+        in context["wires"]. Note: Qubit declarations must be global.
+
+        Args:
+            node (QASMNode): The QubitDeclaration QASMNode.
+            context (dict): The current context.
+        """
+        context["wires"].append(node.qubit.name)
+
+    # needs to have same signature as visit()
+    @visit.register(ClassicalDeclaration)
+    def visit_classical_declaration(
+        self, node: ClassicalDeclaration, context: dict
+    ):  # pylint: disable=no-self-use
+        """
+        Registers a classical declaration. Traces data flow through the context, transforming QASMNodes into Python
+        type variables that can be readily used in expression evaluation, for example.
+
+        Args:
+            node (QASMNode): The ClassicalDeclaration QASMNode.
+            context (dict): The current context.
+        """
+        if node.init_expression is not None:
+            context["vars"][node.identifier.name] = {
+                "ty": node.type.__class__.__name__,
+                "val": node.init_expression.value,
+                "line": node.init_expression.span.start_line,
+            }
+        else:
+            context["vars"][node.identifier.name] = {
+                "ty": node.type.__class__.__name__,
+                "val": None,
+                "line": node.span.start_line,
+            }
+
+    @visit.register(QuantumGate)
+    def visit_quantum_gate(self, node: QuantumGate, context: dict):
+        """
+        Registers a quantum gate application. Calls the appropriate handler based on the sort of gate
+        (parameterized or non-parameterized).
+
+        Args:
+            node (QASMNode): The QuantumGate QASMNode.
+            context (dict): The current context.
+        """
+        name = node.name.name.upper()
+        if name in PARAMETERIZED_GATES:
+            if not node.arguments:
+                raise TypeError(
+                    f"Missing required argument(s) for parameterized gate {node.name.name}"
+                )
+            gates_dict = PARAMETERIZED_GATES
+        elif name in NON_PARAMETERIZED_GATES:
+            gates_dict = NON_PARAMETERIZED_GATES
+        else:
+            raise NotImplementedError(f"Unsupported gate encountered in QASM: {node.name.name}")
+
+        gate, args, wires = self._gate_setup_helper(node, gates_dict, context)
+
+        if len(node.modifiers) > 0:
+            num_control = sum("ctrl" in mod.modifier.name for mod in node.modifiers)
+            op_wires = wires[num_control:]
+            control_wires = wires[:num_control]
+            if "ctrl" in node.modifiers[-1].modifier.name:
+                prev, wires = self.apply_modifier(
+                    node.modifiers[-1], gate(*args, wires=op_wires), context, control_wires
+                )
+            else:
+                prev, wires = self.apply_modifier(
+                    node.modifiers[-1], gate(*args, wires=wires), context, wires
+                )
```

Comment:
I guess I'm still confused as to why we need to do this and why the other loop doesn't work?